### PR TITLE
typings(ApplicationCommandOptionData): Align command data options types more with discord API

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Discord outlines that some combinations aren't valid in application command options.

> `choices?`: choices for STRING, INTEGER, and NUMBER types for the user to pick from
> `options?`: If the option is a subcommand or subcommand group type, this nested options will be the parameters

This PR makes any options not adhering to the api spec cause a compile-time error. I didn't put too much thought into naming some of the types, so alternatives are welcome.

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by advaith1** [comment]
i think this is implemented in discord-api-types, have you looked at that?

**Comment by suneettipirneni** [comment]
> i think this is implemented in discord-api-types, have you looked at that?

So I do see that it's implemented there, however I've never used `discord-api-types`, and I'm unsure how to use it's typings with discord.js's typings, because the types conflict. I'm probably missing something here, but can I use `discord-api-types` with discord.js?

**Review by kyranet — APPROVED** [review]
It'd be good to have some tests in `typings/tests.ts`, but looks good to me :+1:

File: typings/index.d.ts [hunk 1]
```diff
@@ -2889,19 +2889,56 @@ export interface ApplicationCommandData {
...
+export interface ApplicationCommandChoicesData extends BaseApplicationCommandOptionsData {
+  type: CommandOptionChoiceResolvableType;
   choices?: ApplicationCommandOptionChoice[];
-  options?: ApplicationCommandOptionData[];
 }
 
-export interface ApplicationCommandOption extends ApplicationCommandOptionData {
+export interface ApplicationCommandSubOptionsData extends BaseApplicationCommandOptionsData {
+  type: CommandOptionSubOptionResolvableType;
+  options?: this[];
```
Comment (vladfrangu) [inline typings/index.d.ts:2926]: This is inaccurate. In the case of a subcommand group, options is just an "array" of subcommands (usually always has just one value anyways)
In the case of a subcommand, options is any valid option type, excluding subcommand/subcommand group

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: typings/index.d.ts (hunk 1 above)
Comment (suneettipirneni) [reply typings/index.d.ts]: Good catch, this has been fixed.

**Comment by suneettipirneni** [comment]
> It'd be good to have some tests in `typings/tests.ts`, but looks good to me

Will do

**Review by vladfrangu — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Pull request merged**